src/EventStore.Client/EventStoreClientBase.cs

```diff
@@ -32,38 +38,51 @@ public abstract class EventStoreClientBase : IDisposable, IAsyncDisposable {
 		/// <param name="exceptionMap"></param>
 		protected EventStoreClientBase(EventStoreClientSettings? settings,
 			IDictionary<string, Func<RpcException, Exception>> exceptionMap) {
-			_exceptionMap = exceptionMap;
 			Settings = settings ?? new EventStoreClientSettings();
+			_exceptionMap = exceptionMap;
+			_cts = new CancellationTokenSource();
+			_channelSelector = new ChannelSelector(Settings, _cts.Token);
+			_channelInfoLazy = new Lazy<Task<ChannelInfo>>(() => _channelSelector.SelectChannel(_cts.Token),
+				LazyThreadSafetyMode.PublicationOnly);
 
 			ConnectionName = Settings.ConnectionName ?? $"ES-{Guid.NewGuid()}";
-
-			_channels = new MultiChannel(Settings);
 		}
 
-		/// <summary>
-		///
-		/// </summary>
-		/// <param name="cancellationToken"></param>
-		/// <returns></returns>
-		protected async Task<CallInvoker> SelectCallInvoker(CancellationToken cancellationToken) =>
+#pragma warning disable 1591
+		protected Task<ChannelInfo> GetCurrentChannelInfo() => _channelInfoLazy.Value;
+#pragma warning restore 1591
+
+#pragma warning disable 1591
+		protected CallInvoker CreateCallInvoker(ChannelBase channel) =>
 			(Settings.Interceptors ?? Array.Empty<Interceptor>()).Aggregate(
-				(await _channels.GetCurrentChannel(cancellationToken).ConfigureAwait(false)).CreateCallInvoker()
-				.Intercept(new TypedExceptionInterceptor(_exceptionMap))
-				.Intercept(new ConnectionNameInterceptor(ConnectionName))
-				.Intercept(new ReportLeaderInterceptor(_channels.SetEndPoint)),
+				channel.CreateCallInvoker()
+					.Intercept(new TypedExceptionInterceptor(_exceptionMap))
+					.Intercept(new ConnectionNameInterceptor(ConnectionName))
+					.Intercept(new ReportLeaderInterceptor(endPoint => {
+						_channelInfoLazy = new Lazy<Task<ChannelInfo>>(
+							() => _channelSelector.SelectChannel(_cts.Token),
+							LazyThreadSafetyMode.PublicationOnly);
+						_channelSelector.Rediscover(endPoint);
```

Comment:
similarly if two CallInvokers get exceptions handled by their ReportLeaderInterceptors there might be races wrt their calls to the channelSelector